Fix create menu after menu switch

draganescu opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Closes #59604 - create menu is available as an option after menu switching.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Bug fix.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Removes a seemingly useless? call to setting a state about creating a menu (`setIsCreatingMenu`) which makes no sense in the context of switching a menu.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->


1. Have two menus
2. Go to the site editor
3. In a template add a navigation block
4. Open block settings
5. Above the navigation tree click the dot menu in the top right
6. Switch to another menu
7. Open the he dot menu in the top right again
8. Notice "Create new menu" option at the bottom is enabled and works

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->


N/A

## Screenshots or screencast <!-- if applicable -->


https://github.com/WordPress/gutenberg/assets/107534/c8aa4142-08d9-4818-8027-df7e0e51f1d2

draganescu committed:
don't set creating menu flag on menu switch

scruffian reviewed (dismissed):
Makes perfect sense and fixes the issue.

getdave commented:
@scruffian @afercia @draganescu As this bug was uncovered during the RC phase do you feel that it should be included in the WP 6.5? If so why?

Noting that if required, we will need to present a rationale to the release team to clear this to land. As a member of that team I would prefer not to cast my opinion without first receiving your feedback. Much appreciated.

draganescu committed:
return Promises from the navigation menu ref changing functions to be able to clean state after ref is updated

draganescu commented:
In c2c53d1523e40de7f759e5d8198ce19e83c1c155 I've updated the approach to:

- return a promise from the event handlers that alter the ref of the nav menu
- renamed the state to what it really is
- used the resolver to properly update the state.

draganescu committed:
update tests

draganescu committed:
await the menu creation

draganescu committed:
stub test updates

jeryj commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -177,11 +193,16 @@ function NavigationMenuSelector( {
-									setIsCreatingMenu( true );
+									setIsUpdatingMenuRef( true );
+									onCreateNew().then( () => {
+										setIsUpdatingMenuRef( false );
```
Removing this line allows the 'should handle disabled state of the create menu button during the creation process' test to pass. I think it's because the promise onCreateNew is resolving to `false` before the unit test can check for the disabled state.

I believe [this is the line in the unit test](https://github.com/WordPress/gutenberg/pull/59630/files?diff=split&w=1#diff-5acf9be3bfd5e32ba550339fe797b11880fdb50271e88810c2e879ded5f9cddbR278-R282) that both sets the button to be disabled and then resolves to no longer be disabled, so the unit test doesn't even "see" the state of it being disabled.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -342,20 +345,28 @@ function Navigation( {
-				focusNavigationBlock: true,
-			} );
-		}
+		return new Promise( async ( resolve ) => {
```
Async functions _always_ return a `Promise` so this isn't required.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -154,7 +154,10 @@ function Navigation( {
 	} = useCreateNavigationMenu( clientId );
 
 	const createUntitledEmptyNavigationMenu = () => {
-		createNavigationMenu( '' );
+		return new Promise( async ( resolve ) => {
+			await createNavigationMenu( '' );
+			resolve();
+		} );
```
Async functions will always return a `Promise` so you can remove the wrapper and just mark the function as async.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -342,20 +345,28 @@ function Navigation( {
 	};
 
 	const onSelectNavigationMenu = ( menuId ) => {
-		handleUpdateMenu( menuId );
+		return new Promise( ( resolve ) => {
+			handleUpdateMenu( menuId, {
+				focusNavigationBlock: true,
+			} );
+			resolve();
+		} );
```
Why are we using a `Promise` here when there are no async actions occuring?

getdave replied:
`handleUpdateMenu` is sync right?

draganescu replied:
it's not needed indeed, it just liked the exaggerated consistency 😂

draganescu committed:
revert test changes and remove decorative promises

draganescu committed:
do not disable the UI when changing the navigation, disable the UI based on resolving menus

draganescu commented:
I have updated the `NavigationMenuSelector` to NOT disable the menu items when changing the menu. It is a state change in what appears to be a sync operation, so disabling and enabling the UI happens all in the click handler via a local update. If by some magic the same thing is repeated it's just gonna change to the last clicked one.

I've updated the tests accordingly to take into account that the disabled state depends on `hasResolvedNavigationMenus`. 

The more I spend time here the more useless `isUpdatingMenuRef` as a state feels, but as a bug fix this looks ok.

getdave commented on packages/block-library/src/navigation/edit/index.js:
```diff
@@ -355,7 +355,9 @@ function Navigation( {
 
 	const onSelectNavigationMenu = ( menuId ) => {
-		handleUpdateMenu( menuId );
+		handleUpdateMenu( menuId, {
+			focusNavigationBlock: true,
+		} );
```
When testing I found that - as per the code - this focuses the Nav block in the canvas when switching menus.

I wanted to question this, because surely if you are within the list view and you switch menus you'd expect focus to the remain on the menu switching UI? Or at the very least within the editable List View in the block inspector controls?

This change also _appears_ to be outside the scope of this PR. Maybe I missed something but it feels like we should tackle this aspect separately.

draganescu replied:
You're right that it's outside the scope of the PR but I also think it's inconsistent that we focus on the canvas when we create and on the inspector when we switch. I think in both cases we should probably not move focus, it should be returned tot he toggle of the dropdown. I'll remove this change from here.

getdave reviewed (commented):
Thanks for this PR and working on a fix to this bug.

I see the original bug is resolved with the `Create menu` option enabled even when you switch menus.

However, if I click `Create new menu` I would expect the options to switch my menu to be disabled during that process's resolution. Currently they are not for both block and classic menus you might have.

getdave commented on packages/block-library/src/navigation/edit/test/navigation-menu-selector.js:
```diff
@@ -271,6 +271,16 @@ describe( 'NavigationMenuSelector', () => {
 					} )
 				);
 
+				useNavigationMenu.mockReturnValue( {
+					navigationMenus: [],
+					hasResolvedNavigationMenus: false,
+					isResolvingNavigationMenus: true,
+					canUserCreateNavigationMenu: true,
+					canSwitchNavigationMenu: true,
+				} );
```
Nit: I know this is not your fault, but these tests would be a lot easier to read if there were comments above these mocks explaining the state they create. For example: "Set menus to resolving"

draganescu committed:
block the UI for menu switching too, update tests with some comments, undo the focus update

getdave reviewed (commented):
The UI doesn't seem blocked to me when creating a menu. 

If I switch to slow network this becomes very obvious

https://github.com/WordPress/gutenberg/assets/444434/0a72bdf1-892c-4245-ada1-8c1438845b3f

draganescu committed:
properly disable menu choices via their choices prop

getdave reviewed (approved):
Tested and it is now working correctly. I couldn't break it no matter what I triend with slow network or otherwise.

Thanks for your persistence with this problem 👍

getdave commented on packages/block-library/src/navigation/edit/navigation-menu-selector.js:
```diff
@@ -111,17 +121,17 @@ function NavigationMenuSelector( {
 		}
 	}, [
 		hasResolvedNavigationMenus,
 		createNavigationMenuIsSuccess,
 		canUserCreateNavigationMenu,
 		createNavigationMenuIsError,
-		isCreatingMenu,
+		isUpdatingMenuRef,
 		menuUnavailable,
 		noBlockMenus,
 		noMenuSelected,
```
We should be able to remove all the unused deps from this effect. 

I propose we do this in a _followup_.

getdave commented:
I just manually cherry-picked this PR to the pick/wp-65-rc-2 branch to get it included in the next release

https://github.com/WordPress/gutenberg/pull/59756

afercia commented:
While this PR fixes the disabled state of the 'Create new menu' button, it introduces a regression. There is now a focus loss when creating a new menu because the whole dropdown menu stays open but all menu items gets 'disabled', including the menu item that had keyboard focus.

It is important to point out, once again, a very important principle:

> [!IMPORTANT]
> Never, ever, disable controls dynamically while they have focus.

I'd appreciate that this kind of changes are _tested with keyboard_ as clearly this wasn't the case in this PR. Thanks.

I will create a separate issue, and this needs to be addressed soon as it's a regression in RC.

draganescu commented:
@afercia how was this a regression? Isn't it more like ... the bug being fixed unearthed the existing problem?

getdave commented:
> and this needs to be addressed soon as it's a regression in RC.

Agreed. Are you able to tackle this one? 

The next RC is RC3 on 19th March.
